x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/preprocessor/FullTextFunctionMapperPreprocessor.java

```diff
@@ -0,0 +1,129 @@
+import org.elasticsearch.xpack.esql.expression.function.fulltext.FullTextFunction;
+import org.elasticsearch.xpack.esql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.planner.TranslatorHandler;
+import org.elasticsearch.xpack.esql.plugin.TransportActionServices;
+import org.elasticsearch.xpack.esql.session.IndexResolver;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Some {@link FullTextFunction} implementations such as {@link org.elasticsearch.xpack.esql.expression.function.fulltext.Match}
+ * will be translated to a {@link QueryBuilder} that require a rewrite phase on the coordinator.
+ * {@link FullTextFunctionMapperPreprocessor#preprocess(LogicalPlan, TransportActionServices, ActionListener)} will rewrite the plan by
+ * replacing {@link FullTextFunction} expression with new ones that hold rewritten {@link QueryBuilder}s.
+ */
+public class FullTextFunctionMapperPreprocessor implements MappingPreProcessor {
+
+    @Override
+    public void preprocess(LogicalPlan plan, TransportActionServices services, ActionListener<LogicalPlan> listener) {
+        Set<FullTextFunction> unresolved = fullTextFunctions(plan);
+        Set<String> indexNames = indexNames(plan);
+
+        if (indexNames == null || indexNames.isEmpty() || unresolved.isEmpty()) {
+            listener.onResponse(plan);
+            return;
+        }
+        QueryRewriteContext ctx = queryRewriteContext(services, indexNames);
+        FullTextFunctionsRewritable rewritable = new FullTextFunctionsRewritable(unresolved);
+        Rewriteable.rewriteAndFetch(
+            rewritable,
+            ctx,
+            listener.delegateFailureAndWrap((l, r) -> l.onResponse(updateQueryBuilders(plan, r.results())))
+        );
+    }
+
+    public LogicalPlan updateQueryBuilders(LogicalPlan plan, Map<FullTextFunction, QueryBuilder> newQueryBuilders) {
+        LogicalPlan newPlan = plan.transformExpressionsDown(FullTextFunction.class, m -> {
+            if (newQueryBuilders.containsKey(m)) {
+                return m.replaceQueryBuilder(newQueryBuilders.get(m));
+            }
+            return m;
+        });
+        // The given plan was already analyzed and optimized, so we set the resulted plan to optimized as well.
+        newPlan.setOptimized();
+        return newPlan;
+    }
+
+    private static QueryRewriteContext queryRewriteContext(TransportActionServices services, Set<String> indexNames) {
+        ResolvedIndices resolvedIndices = ResolvedIndices.resolveWithIndexNamesAndOptions(
+            indexNames.toArray(String[]::new),
+            IndexResolver.FIELD_CAPS_INDICES_OPTIONS,
+            services.clusterService().state(),
+            services.indexNameExpressionResolver(),
+            services.transportService().getRemoteClusterService(),
+            System.currentTimeMillis()
+        );
```

Comment:
The indices have already been resolved through field-caps; double check if this information can be reused instead of going back to string format.

Reply:
`ResolvedIndices` splits the indices by local and remote and uses the string format for that. Not sure if/how we can simplify this.